src/jit/morph.cpp

```diff
@@ -4963,11 +4963,18 @@ GenTreePtr Compiler::fgMorphMultiregStructArg(GenTreePtr arg, fgArgTabEntryPtr f
         objClass           = argObj->gtClass;
         structSize         = info.compCompHnd->getClassSize(objClass);
 
-        // If we have a GT_OBJ of a GT_ADDR then we set argValue to the child node of the GT_ADDR
-        //
-        if (argObj->gtOp1->OperGet() == GT_ADDR)
+        // If we have a GT_OBJ of a GT_ADDR then we set argValue to the child node of the GT_ADDR.
+        GenTree* op1 = argObj->gtOp1;
+        if (op1->OperGet() == GT_ADDR)
         {
-            argValue = argObj->gtOp1->gtOp.gtOp1;
+            GenTree* underlyingTree = op1->gtOp.gtOp1;
+
+            // Only update to the same type.
+            if (underlyingTree->TypeGet() == argValue->TypeGet())
+            {
+                argValue = underlyingTree;
+                assert(objClass == gtGetStructHandleIfPresent(underlyingTree));
```

Comment:
I think it would be possible in IL to take the address of a lclVar, and then use ldobj with a different struct type (including size) to pass it to a method. While it's not verifiable IL, I don't think it is illegal. In that case, this assert would fail. Also, the arg setup code, such as in `fgMorphMultiregStructArg()` would use the strong struct layout for passing the struct.
I think you need to turn the above assert into the condition for eliding the OBJ(ADDR())